### ISSUE TITLE
Kopia v0.10 with velero 1.10.2 high memory usage when opening large repo for maintenance

### ISSUE THREAD

**Issue by ktogias** [kopia/kopia#2888]
We have velero+kopia on a kuberentes cluster for cluster+pv back up. 

Velero pod consumes large amounts of memory when running maintainance on an ns backup repository with 2 large pvs (about 250GB each) and several smaller (16-64GB each), it exceeds the limit of 5.5GB of RAM we have set and gets OOM killed. This happens repeatedly and the pod enters CrashLoopbackOff state.

Trying to increase the limit causes the node running the velero pod getting not ready and needing restart (the node has 16GB RAM total).

Deleting all the backups and repository of the specific namespace solves temporary the problem, but if we start again to take regular backups of the specific namespace, we end on OOM after some days. We have tryied velero latest version 1.10.2 with no luck.

The problem seems to happen when opening the backup repository that is too big to fit in our node available memory for maintainance.

The backup repository is hosted on Minio, its size is 1.1TB and contains 166340 files (find  -type f | wc -l).

I attach velero pod debug logs: 
[logs-from-velero-in-velero-857d759cfb-t4vbx.log](https://github.com/kopia/kopia/files/11147651/logs-from-velero-in-velero-857d759cfb-t4vbx.log)
[logs-from-velero-in-velero-857d759cfb-t4vbx-1.log](https://github.com/kopia/kopia/files/11147653/logs-from-velero-in-velero-857d759cfb-t4vbx-1.log)

**Comment by Lyndon-Li** [comment]
From the log, we can see the problematic repo is running index compact:

```
time="2023-03-31T14:03:13Z" level=debug msg="Creating cache directory '/home/nonroot/.cache/kopia/7919d759145236c6' with max size 2097152000"
time="2023-03-31T14:03:13Z" level=debug msg="kopia.repository could not be fetched from cache: unable to open cache file: stat /home/nonroot/.cache/kopia/7919d759145236c6/kopia.repository: no such file or directory"
time="2023-03-31T14:03:13Z" level=debug msg="kopia.blobcfg could not be fetched from cache: unable to open cache file: stat /home/nonroot/.cache/kopia/7919d759145236c6/kopia.blobcfg: no such file or directory"
time="2023-03-31T14:03:13Z" level=debug msg="throttling limits from connection info" limits="{0 0 0 0 0}"

time="2023-03-31T14:03:13Z" level=debug msg="refreshing committed state because it's no longer valid (now 2023-03-31 14:03:13.896558729 +0000 UTC, valid until 0001-01-01T00:00:00Z)"
time="2023-03-31T14:03:13Z" level=debug msg=refreshAttemptLocked
time="2023-03-31T14:03:14Z" level=debug msg="ranges: []"
time="2023-03-31T14:03:14Z" level=debug msg="current epoch 9, uncompacted epoch sets 20 0 0, valid until 2023-03-31T14:23:13.897010906Z"
time="2023-03-31T14:03:14Z" level=debug msg="generating range checkpoint"
time="2023-03-31T14:03:14Z" level=debug msg="adding incremental state for epochs 0..10 on top of []"
time="2023-03-31T14:03:14Z" level=debug msg="generating range checkpoint for 0..7"
time="2023-03-31T14:03:14Z" level=debug msg="adding incremental state for epochs 0..7 on top of []"
time="2023-03-31T14:03:14Z" level=debug msg="Complete Index Set for [0..10]: [xs0_56cb8a340f35d629ffe22f4df77f5953-s28da87f8d88beae3-c1 xs1_8c9132863a466e65dded293f3353b919-sadffd632911df667-c1 xs2_fb3de18d6b7e88d458e0a2d73f5617dc-sb83795d242e03d40-c1 xs3_0b8c366e57952055c7e5ef26035fa2c7-s6e63f6b8acc00d8a-c1 xs4_def2c5c740e3e5f0080ae7cd87cff422-s57bed458b7851ba1-c1 xs5_d2fa439a7445e77a47fb6205f262090c-sbd6163b72789d315-c1 xs6_f6ce050019668073036394d5a77a4657-s64f8c11b66f1d22c-c1 xs7_4ff5d750f1c1dd7e4ae680b1044422fd-s5130f4deab4b7127-c1 xn8_0442d9f1320e4c9835714c31f6d80b22-s1e0256280e23845b11a-c1 xn8_13e8bb87f00306d8094aebf5b37f026b-s6924d58f3ad6035b11a-c1 xn8_169b650a6bf86dfc6e2de6c0f59477dd-sce09fa1f6652f7b411a-c1 xn8_1f459fa5abe839df85675295cd3af837-sd9266dbabfe75efe11a-c1 xn8_2220e674ace93665351a0059e1e71849-s8aadb51feeeeea5811a-c1 xn8_4ddb9e258540cd2498f7ae10b52930b7-s0a2c382e3328cf5411a-c1 xn8_53ef03ac7869cb5b0ac73d3b07fcadee-s1fc938ecc751894011a-c1 xn8_606180f3c5628930ddcd7380cc82fe44-s8ab66cb4b30fd4a211a-c1 xn8_63f910b842630688a715e5b2c6db3605-sd8c69f2655c6be6f11a-c1 xn8_91904c45c2b1f885d0ad184585c47d6d-saa64f7ca5503c30211a-c1 xn8_93cd8b2bb4d6391ac54f00558bfab3a6-s67cb0008696343b811a-c1 xn8_93d9e297ef651aa3e1ba4b947ce68c4d-s9fa1f1d3caafbc0011a-c1 xn8_a635ec69e9bf515225ce552fd6ac896e-sce6a3601d592756811a-c1 xn8_c6ea98ddc20d1edab0b97306efdb5d20-scf6c681640926b3a11a-c1 xn8_c876ddd4990b0e0f47b1c755038fe371-sb39000599f9256af11a-c1 xn8_d44bb6f9e9c49e02d18974b2d54437bd-sd63e3f0d4f15438211a-c1 xn8_e5fd2269ea1047e16457c7da738c2555-sfa9374f4b1234ae711a-c1 xn8_eee984c1d2f0288998342d0fa460c65d-s115aca13aa4d24cc11a-c1 xn8_f79e195d1f667468958a63d4b5c31ebc-s856e8a2575f5e15e11a-c1 xn8_fd3d90a78ccd0b24a88768fde75e3394-s6881f083a854307611a-c1], deletion watermark 0001-01-01 00:00:00 +0000 UTC"
time="2023-03-31T14:03:14Z" level=warning msg="active indexes [xs0_56cb8a340f35d629ffe22f4df77f5953-s28da87f8d88beae3-c1 xs1_8c9132863a466e65dded293f3353b919-sadffd632911df667-c1 xs2_fb3de18d6b7e88d458e0a2d73f5617dc-sb83795d242e03d40-c1 xs3_0b8c366e57952055c7e5ef26035fa2c7-s6e63f6b8acc00d8a-c1 xs4_def2c5c740e3e5f0080ae7cd87cff422-s57bed458b7851ba1-c1 xs5_d2fa439a7445e77a47fb6205f262090c-sbd6163b72789d315-c1 xs6_f6ce050019668073036394d5a77a4657-s64f8c11b66f1d22c-c1 xs7_4ff5d750f1c1dd7e4ae680b1044422fd-s5130f4deab4b7127-c1 xn8_0442d9f1320e4c9835714c31f6d80b22-s1e0256280e23845b11a-c1 xn8_13e8bb87f00306d8094aebf5b37f026b-s6924d58f3ad6035b11a-c1 xn8_169b650a6bf86dfc6e2de6c0f59477dd-sce09fa1f6652f7b411a-c1 xn8_1f459fa5abe839df85675295cd3af837-sd9266dbabfe75efe11a-c1 xn8_2220e674ace93665351a0059e1e71849-s8aadb51feeeeea5811a-c1 xn8_4ddb9e258540cd2498f7ae10b52930b7-s0a2c382e3328cf5411a-c1 xn8_53ef03ac7869cb5b0ac73d3b07fcadee-s1fc938ecc751894011a-c1 xn8_606180f3c5628930ddcd7380cc82fe44-s8ab66cb4b30fd4a211a-c1 xn8_63f910b842630688a715e5b2c6db3605-sd8c69f2655c6be6f11a-c1 xn8_91904c45c2b1f885d0ad184585c47d6d-saa64f7ca5503c30211a-c1 xn8_93cd8b2bb4d6391ac54f00558bfab3a6-s67cb0008696343b811a-c1 xn8_93d9e297ef651aa3e1ba4b947ce68c4d-s9fa1f1d3caafbc0011a-c1 xn8_a635ec69e9bf515225ce552fd6ac896e-sce6a3601d592756811a-c1 xn8_c6ea98ddc20d1edab0b97306efdb5d20-scf6c681640926b3a11a-c1 xn8_c876ddd4990b0e0f47b1c755038fe371-sb39000599f9256af11a-c1 xn8_d44bb6f9e9c49e02d18974b2d54437bd-sd63e3f0d4f15438211a-c1 xn8_e5fd2269ea1047e16457c7da738c2555-sfa9374f4b1234ae711a-c1 xn8_eee984c1d2f0288998342d0fa460c65d-s115aca13aa4d24cc11a-c1 xn8_f79e195d1f667468958a63d4b5c31ebc-s856e8a2575f5e15e11a-c1 xn8_fd3d90a78ccd0b24a88768fde75e3394-s6881f083a854307611a-c1] deletion watermark 0001-01-01 00:00:00 +0000 UTC"
time="2023-03-31T14:03:14Z" level=debug msg="Downloading 28 new index blobs..."
time="2023-03-31T14:03:36Z" level=debug msg="Index blobs downloaded."
time="2023-03-31T14:03:36Z" level=debug msg="use-indexes [xs0_56cb8a340f35d629ffe22f4df77f5953-s28da87f8d88beae3-c1 xs1_8c9132863a466e65dded293f3353b919-sadffd632911df667-c1 xs2_fb3de18d6b7e88d458e0a2d73f5617dc-sb83795d242e03d40-c1 xs3_0b8c366e57952055c7e5ef26035fa2c7-s6e63f6b8acc00d8a-c1 xs4_def2c5c740e3e5f0080ae7cd87cff422-s57bed458b7851ba1-c1 xs5_d2fa439a7445e77a47fb6205f262090c-sbd6163b72789d315-c1 xs6_f6ce050019668073036394d5a77a4657-s64f8c11b66f1d22c-c1 xs7_4ff5d750f1c1dd7e4ae680b1044422fd-s5130f4deab4b7127-c1 xn8_0442d9f1320e4c9835714c31f6d80b22-s1e0256280e23845b11a-c1 xn8_13e8bb87f00306d8094aebf5b37f026b-s6924d58f3ad6035b11a-c1 xn8_169b650a6bf86dfc6e2de6c0f59477dd-sce09fa1f6652f7b411a-c1 xn8_1f459fa5abe839df85675295cd3af837-sd9266dbabfe75efe11a-c1 xn8_2220e674ace93665351a0059e1e71849-s8aadb51feeeeea5811a-c1 xn8_4ddb9e258540cd2498f7ae10b52930b7-s0a2c382e3328cf5411a-c1 xn8_53ef03ac7869cb5b0ac73d3b07fcadee-s1fc938ecc751894011a-c1 xn8_606180f3c5628930ddcd7380cc82fe44-s8ab66cb4b30fd4a211a-c1 xn8_63f910b842630688a715e5b2c6db3605-sd8c69f2655c6be6f11a-c1 xn8_91904c45c2b1f885d0ad184585c47d6d-saa64f7ca5503c30211a-c1 xn8_93cd8b2bb4d6391ac54f00558bfab3a6-s67cb0008696343b811a-c1 xn8_93d9e297ef651aa3e1ba4b947ce68c4d-s9fa1f1d3caafbc0011a-c1 xn8_a635ec69e9bf515225ce552fd6ac896e-sce6a3601d592756811a-c1 xn8_c6ea98ddc20d1edab0b97306efdb5d20-scf6c681640926b3a11a-c1 xn8_c876ddd4990b0e0f47b1c755038fe371-sb39000599f9256af11a-c1 xn8_d44bb6f9e9c49e02d18974b2d54437bd-sd63e3f0d4f15438211a-c1 xn8_e5fd2269ea1047e16457c7da738c2555-sfa9374f4b1234ae711a-c1 xn8_eee984c1d2f0288998342d0fa460c65d-s115aca13aa4d24cc11a-c1 xn8_f79e195d1f667468958a63d4b5c31ebc-s856e8a2575f5e15e11a-c1 xn8_fd3d90a78ccd0b24a88768fde75e3394-s6881f083a854307611a-c1]"
time="2023-03-31T14:03:36Z" level=debug msg="combined 28 into 25 index segments"
time="2023-03-31T14:03:36Z" level=debug msg=expireUnused except="[xs0_56cb8a340f35d629ffe22f4df77f5953-s28da87f8d88beae3-c1 xs1_8c9132863a466e65dded293f3353b919-sadffd632911df667-c1 xs2_fb3de18d6b7e88d458e0a2d73f5617dc-sb83795d242e03d40-c1 xs3_0b8c366e57952055c7e5ef26035fa2c7-s6e63f6b8acc00d8a-c1 xs4_def2c5c740e3e5f0080ae7cd87cff422-s57bed458b7851ba1-c1 xs5_d2fa439a7445e77a47fb6205f262090c-sbd6163b72789d315-c1 xs6_f6ce050019668073036394d5a77a4657-s64f8c11b66f1d22c-c1 xs7_4ff5d750f1c1dd7e4ae680b1044422fd-s5130f4deab4b7127-c1 xn8_0442d9f1320e4c9835714c31f6d80b22-s1e0256280e23845b11a-c1 xn8_13e8bb87f00306d8094aebf5b37f026b-s6924d58f3ad6035b11a-c1 xn8_169b650a6bf86dfc6e2de6c0f59477dd-sce09fa1f6652f7b411a-c1 xn8_1f459fa5abe839df85675295cd3af837-sd9266dbabfe75efe11a-c1 xn8_2220e674ace93665351a0059e1e71849-s8aadb51feeeeea5811a-c1 xn8_4ddb9e258540cd2498f7ae10b52930b7-s0a2c382e3328cf5411a-c1 xn8_53ef03ac7869cb5b0ac73d3b07fcadee-s1fc938ecc751894011a-c1 xn8_606180f3c5628930ddcd7380cc82fe44-s8ab66cb4b30fd4a211a-c1 xn8_63f910b842630688a715e5b2c6db3605-sd8c69f2655c6be6f11a-c1 xn8_91904c45c2b1f885d0ad184585c47d6d-saa64f7ca5503c30211a-c1 xn8_93cd8b2bb4d6391ac54f00558bfab3a6-s67cb0008696343b811a-c1 xn8_93d9e297ef651aa3e1ba4b947ce68c4d-s9fa1f1d3caafbc0011a-c1 xn8_a635ec69e9bf515225ce552fd6ac896e-sce6a3601d592756811a-c1 xn8_c6ea98ddc20d1edab0b97306efdb5d20-scf6c681640926b3a11a-c1 xn8_c876ddd4990b0e0f47b1c755038fe371-sb39000599f9256af11a-c1 xn8_d44bb6f9e9c49e02d18974b2d54437bd-sd63e3f0d4f15438211a-c1 xn8_e5fd2269ea1047e16457c7da738c2555-sfa9374f4b1234ae711a-c1 xn8_eee984c1d2f0288998342d0fa460c65d-s115aca13aa4d24cc11a-c1 xn8_f79e195d1f667468958a63d4b5c31ebc-s856e8a2575f5e15e11a-c1 xn8_fd3d90a78ccd0b24a88768fde75e3394-s6881f083a854307611a-c1]" minSweepAge=1h0m0s

time="2023-03-31T14:03:36Z" level=debug msg="closing shared manager"
```

Is the memory usage expected for this operation?

**Comment by julio-lopez** [comment]
Can you reproduce this issue with the most recent version of kopia?

**Comment by Lyndon-Li** [comment]
@julio-lopez there are some difficulties to reproduce it with the latest version because Velero is current shipping Kopia v0.10.7, but let me try with @ktogias, this may need some time.

**Comment by Lyndon-Li** [comment]
@julio-lopez 
Meanwhile, I have compared the code & release notes between v0.10.7 and v0.12.1 and not found significant changes around the epoch manager and index compacting, do you know any?
Or is the problem not relevant to index compact?

**Comment by julio-lopez** [comment]
I do not know at the top of my head whether or not there have been relevant changes since 10.7. That was release more than a year ago.

At first, it looks related to index compaction, but it could be a side effect of another issue.
While "full" maintenance can be memory intensive, index compaction is not expected to be so.

It'd be good to reproduce with a recent version of kopia, ideally from `HEAD`.

@ktogias Can you include the output of these commands:

```
kopia content stats
kopia blob stats
kopia maintenance info
```

**Comment by Lyndon-Li** [comment]
@julio-lopez 
> While "full" maintenance can be memory intensive

At beginning I also thought it was related to maintenance, but after checking the log back and force, I found the maintenance had not started at all. The story was:

- Velero found the problematic repo on startup
- it wanted to check the healthy of the connection to the repo, so it called `Repository.Connect` 
- In the implementation of `Repository.Connect`, the repo is connected then opened for verification by `verifyConnect`
- The verification went well, so the repo is closed
- Looks like the open of the repo triggered some async operation (I guess it was index compaction), so the close operation waited there until oom

It means the `Connect` call had not returned, so maintenance was not started.

**Comment by julio-lopez** [comment]
Got it, thanks for the context.

Content stats and index sizes would be useful here

**Comment by ktogias** [comment]
Hi all,

@julio-lopez 
I managed to manually connect to the affected repository and run the suggested commands through my desktop using kopia docker image. When I run kopia image with no memory limit, it consumed all the available ram and swap of my 16GB RAM+16GB swap desktop causing it to become unresponsive. 

What I did:

`docker run --memory 8g --name kopia kopia/kopia:0.10 server start --log-level=debug --insecure`

[server.log](https://github.com/kopia/kopia/files/11200317/server.log)

`docker exec kopia ./kopia repository connect s3 --endpoint=<>:9000 --bucket=vkub --prefix=kopia/<>/ --access-key=<> --secret-access-key=<> --password=<> --log-level=debug --readonly --disable-tls `

[connect.log](https://github.com/kopia/kopia/files/11200337/connect.log)

`docker exec kopia ./kopia content stats --log-level=debug --password=<>`

[content-stats.log](https://github.com/kopia/kopia/files/11200345/content-stats.log)

`docker exec kopia ./kopia blob stats --log-level=debug --password=<>`

[blob-stats.log](https://github.com/kopia/kopia/files/11200349/blob-stats.log)

`docker exec kopia ./kopia index list --json --log-level=debug --password=<>`

[index-list.log](https://github.com/kopia/kopia/files/11200365/index-list.log)

`docker exec kopia ./kopia maintenance info --json --log-level=debug --password=<>`

[maintenance-info.log](https://github.com/kopia/kopia/files/11200378/maintenance-info.log)

Log files attached after each command contain the output of the command.

I have tested with v0.10 of kopia. 

I will be glad to run any other commands or tests on the repo with v0.10 or later kopia versions and provide you the output, or any other info.

**Comment by Shrekster** [comment]
Attempted a basic repro with v0.10.7 binary, flat dir, 1 million 100k files and ran full maintenance. (I did not reduce the maintenance safety while doing this). Target is minio.

![plot](https://user-images.githubusercontent.com/1415030/231875727-26ee5a77-467b-48df-b52f-1716b37004db.svg)

I'll attempt again in a bit with bumping the source to 1TB (still 1mil files) and reducing maintenance safety which will  force it do blob deletions.

**Comment by Lyndon-Li** [comment]
@Shrekster
As the discussion here https://github.com/kopia/kopia/issues/2888#issuecomment-1502630503, the problem didn't happen during maintenance, but happened during open & close repo.

**Comment by julio-lopez** [comment]
@ktogias Thanks for the info.

A couple of high-level stats
- The size of the index files is ~780MB total.
- There are about 20+ million unique contents in the indexes.

You mentioned there are many small files in the repo. Does that match the expectation, as it is in the order of ~20 million total?
Have snapshots been deleted from the repository? Wondering if there are contents in the repo that are no longer in use by snapshots.

I discussed the issue with @Shrekster earlier in the week and had not updated things here yet. He's currently attempting to reproduce the issue.

Thanks

**Comment by Shrekster** [comment]
Thanks @Lyndon-Li . I will scale up the repo to make sure I get 20mil index entries. Also, will need to figure out a way to introduce index companions. (I think a relatively large number of snapshots + deletes would cause that).

> The size of the index files is ~780MB total.

Ok. In my test the index size is 250MB, I'll scale that up as well.

**Comment by ashmrtn** [comment]
I'm not familiar with how the older index works, but just a heads up that I've seen high memory usage from golang's stdlib JSON decoder when deserializing large sets of snapshot manifests. The underlying issue is that the stdlib JSON decoder is not particularly memory efficient when deserializing large arrays of objects. If the old index also uses JSON to store data that may be something to look into

I've opened #2956 which has some manual parsing of snapshot manifests to help reduce memory usage a bit. A solution similar to that may work if the old index is parsing lots of JSON arrays as well

**Comment by Shrekster** [comment]
@Lyndon-Li @ktogias How large is the snapshot chain on your repo ? `kopia snapshot list --all`
Have you been deleting snapshots ?

**Comment by Lyndon-Li** [comment]
@ktogias Thanks for the information.
Looks like there are 10 pods in the namespace and each of which has 2 PVs.
Looks like the backup were created daily and most of the snapshot were created around 03-20, so the older snapshot should have been delete successfully

cc @Shrekster

**Comment by ktogias** [comment]
Hi all, 

I increased the maintenanceFrequency of the BackupRepository resource to 2160h (just something big enough to not fire the maintainance). I think that now it get OOM killed when tryinig to open the repo for deleting expired backups. I am attaching a recent log in case it provides any further insight.

[logs-from-velero-in-velero-84f99cc9dc-9wvxp.log](https://github.com/kopia/kopia/files/11310912/logs-from-velero-in-velero-84f99cc9dc-9wvxp.log)

**Comment by Shrekster** [comment]
In my repro attempts I increased the epoch refresh frequency so that we'll cause a lot more compaction:

```
kopia repo set-parameters --epoch-min-duration=10m --epoch-refresh-frequency=3m --max-pack-size-mb=10
```

Also disabled any global policy from kicking in deletes:

```
kopia policy set --global --keep-latest 0 --keep-hourly 0 --keep-daily 0 --keep-weekly 0 --keep-monthly 0 --keep-annual 0 --snapshot-interval 0 --max-parallel-file-reads 300
```

And disabled both full and quick maintenance because I intend to run it the nI am ready for repro:

```
kopia maintenance set --enable-full 0 --enable-quick 0
```

I've slightly less than 60mil objects in my repository of 6TB:

```
Count: 58039599
Total Bytes: 6 TB
Total Packed: 5.9 TB (compression 0.1%)
By Method:
  (uncompressed)         count: 58035835 size: 5.9 TB
  zstd-fastest           count: 3764 size: 10.1 GB packed: 1.3 GB compression: 86.7%
Average: 102.6 KB
Histogram:

        0 between 0 B and 10 B (total 0 B)
        0 between 10 B and 100 B (total 0 B)
      688 between 100 B and 1 KB (total 138.2 KB)
      420 between 1 KB and 10 KB (total 2.5 MB)
     2648 between 10 KB and 100 KB (total 106.4 MB)
 58035827 between 100 KB and 1 MB (total 5.9 TB)
       16 between 1 MB and 10 MB (total 17.3 MB)
        0 between 10 MB and 100 MB (total 0 B)
```

I developed a snapshot chain of 795 after which I started deleting random snapshots to maintain that number:

```
kopia snapshot list --all | wc -l
795
```

After 795 snapshots I started deleting one snapshot for every one new snapshot:

```
kopia snapshot delete --delete $(kopia snapshot list --json | jq -r .[].id | shuf -n 1)
```

As is evident from the settings above, I've not run maintenance (similar to the velcro case):

```
kopia maintenance info
Owner: home@home12
Quick Cycle:
  scheduled: false
Full Cycle:
  scheduled: false
```

My repository has accumulated about half a million blobs:

```
kopia blob stats
Count: 571836
Total: 6 TB
Average: 10.4 MB
Histogram:

        0 between 0 B and 10 B (total 0 B)
        3 between 10 B and 100 B (total 54 B)
      697 between 100 B and 1 KB (total 101 KB)
     1547 between 1 KB and 10 KB (total 6.7 MB)
       50 between 10 KB and 100 KB (total 2.8 MB)
     1555 between 100 KB and 1 MB (total 1.3 GB)
     4485 between 1 MB and 10 MB (total 8.5 GB)
   563499 between 10 MB and 100 MB (total 5.9 TB)
```

The repository index is about 5.8G in size:

```
kopia index list
xr0_7_07ed0e43bdbf16448aff39b0ab926987-s81c4edff909d3764-c1    86909459 2023-04-22 01:55:49.748 PDT []
xr8_15_35df67e6e3716a3114d8fc42b3a7fa56-s2bf6ec13bbcd1da2-c1   86434507 2023-04-22 03:35:11.734 PDT []
xr16_23_f7505d491eaf5b036899632926af0f9a-s6563e294d8a54ca3-c1   86433890 2023-04-22 05:16:53.839 PDT []
xr24_31_fa1a6479d637e8c7c01cef79f6e60b0c-s6210603b8e327ad9-c1   86434387 2023-04-22 07:00:22.280 PDT []
xr32_39_7f40fb1269a78ba90b13de0b43a82623-s46802d8f8fa3a8da-c1   86434450 2023-04-22 08:48:34.570 PDT []
xr40_47_8b0f6e92e80f43245ba41b26409f2dec-s5086202fbf402d38-c1   86471248 2023-04-22 10:38:25.038 PDT []
xr48_55_98b4fae700b2185f9840f4ae8b916a89-s8714292d0c06b168-c1   86455725 2023-04-22 12:29:49.832 PDT []
xr56_63_1d58a1cab093267d84223734cfb6325e-s1cd6a85cd55cb98d-c1   86456970 2023-04-22 14:29:09.651 PDT []
xr64_71_fb35babf1acaeb998b0e0bd9f8811990-s47ba529fd72c8bb7-c1   86456072 2023-04-22 16:39:49.897 PDT []
xr72_79_3f3f36b9e31144366ddda82ad51cd1f6-sb6bb5856f35638b2-c1   86455795 2023-04-22 18:36:51.714 PDT []
xr80_87_020e8cdfcb5e366032c9953a82897652-s58af4b0fc1d1400e-c1   86455893 2023-04-22 20:35:46.535 PDT []
xr88_95_db29962040713fb18e82605cb494565e-s5991853738c10337-c1   86455684 2023-04-22 22:37:26.496 PDT []
xr96_103_8c2e754b661b7aee4764a7e8058241c7-s2ee3a98ead9c02c7-c1   86848467 2023-04-23 00:43:06.076 PDT []
xr104_111_b60d56164fe22772ad352fa6c7df0f36-se6d0b935b43ab8a0-c1   86462759 2023-04-23 02:53:54.155 PDT []
xr112_119_0bbc57b8e9efc17d080160fe54e26d59-sc0078316c5087c96-c1   86462843 2023-04-23 05:01:57.634 PDT []
xr120_127_93d17284d45907da353d4249f12b4e59-s6bfeb227e95b1a4f-c1   86463124 2023-04-23 07:10:44.507 PDT []
xr128_135_6fa678d754bbaff810b4a2ca88bcb60a-se3ba72b5ae5a71ba-c1   86462775 2023-04-23 09:21:13.254 PDT []
xr136_143_a0219fb9455d5e878d6ca207b7276e68-s38b4ebf4c35192d2-c1   86463412 2023-04-23 11:32:37.943 PDT []
xr144_151_64e2a26bffda1942ee0af2ea12b04277-s3c7752022fcf8ab1-c1   86461999 2023-04-23 13:49:05.827 PDT []
xr152_159_8aee4035340688f296c21578b37e0227-s69485ef8d14ed600-c1   86461974 2023-04-23 16:49:55.015 PDT []
xr160_167_2810ef23f05620755ece5a3353afe371-se3b30197a4c0d6bd-c1   86461455 2023-04-23 19:49:02.997 PDT []
xr168_175_6cc997722225f5f591fe2576dc14d84a-s1b9c314abf2798e0-c1   86461965 2023-04-23 22:37:08.500 PDT []
xn184_3a2d8a7948a1e2edf93a6853efb0c0a0-s9a49dc42c311dc3311b-c1        143 2023-04-24 01:32:48.469 PDT []
xn184_cd33599a34f8494f68aac1d1b8272145-sfb53a9368c29183f11b-c1    1350950 2023-04-24 01:41:08.340 PDT []
xn184_737607d44c1f662fb619b2424eb89a25-s8b20f09cc84d42a811b-c1        143 2023-04-24 01:41:10.052 PDT []
xn184_638427a4366ffbf6567f00f74d926975-s59cb54bf52fa10d611b-c1    1350882 2023-04-24 01:48:43.417 PDT []
xn184_1389425fae2a84f7cc180b43410c7ee1-s11dfe3cae25d818511b-c1        143 2023-04-24 01:48:44.925 PDT []
xn184_528ed8c9b5e5d09f4527e5b92f328b84-s64db68e16646e66611b-c1    1350882 2023-04-24 01:54:52.643 PDT []
xn184_23858fc42228ae4478731f3c18a7fc3b-s78f22479d7dcd04511b-c1        143 2023-04-24 01:54:54.063 PDT []
xn184_7e1b4f57f9640836f688d31d64b93279-s8f420ab791d0719811b-c1    1350950 2023-04-24 02:03:10.549 PDT []
xn184_24862c526bcc028025bd031a23aadc76-sec718ad372d92ef511b-c1        143 2023-04-24 02:03:11.554 PDT []
xn184_b0860a8b6174f7e75d90e34c4d725f50-s034d1df8271b615c11b-c1    1350950 2023-04-24 02:11:14.888 PDT []
xn184_1739f63b45283ddb0f724f1383607b73-s006704fcc31c7d6611b-c1        143 2023-04-24 02:11:16.704 PDT []
xn184_26804c140525c6a1037911620ff312cd-s8222139863d2c3e411b-c1    1350950 2023-04-24 02:19:32.387 PDT []
xn184_d17a180516ede72c687414d70d9eb5ee-sd3a3cd7947bac5ce11b-c1       1615 2023-04-24 02:19:32.563 PDT []
xn184_0b747abdd9b49c8d08ba3a84d193c120-s76dc07ed967a5a8211b-c1        143 2023-04-24 02:19:33.967 PDT []
xn184_e7fd91a6695013b966b35b691dc57955-s0a7d38ce0c7f3f5e11b-c1    1350882 2023-04-24 02:24:05.003 PDT []
xn184_2f572442f96ad38fd040835be997b28d-s3e056c2e4acee69d11b-c1        143 2023-04-24 02:24:06.663 PDT []
xn184_1909b9b2080c8d57ba4c4d36ba9516c1-sbeded6d2b146a22e11b-c1    1350950 2023-04-24 02:28:20.847 PDT []
xr176_183_e1a223a7f9fdef083eb8a2dccabb1e41-s4b922d399b451239-c1   86461829 2023-04-24 02:28:33.667 PDT []
xn185_68c295fff00b87a4ca362cec7bc6d0cb-s3f472a5e0ca1c76611b-c1        143 2023-04-24 02:28:35.811 PDT []
xn185_0c401f9c482ea54fe162d8f4bc6cad10-s5e87b4cc068f790111b-c1    1350882 2023-04-24 02:37:20.127 PDT []
xn185_4f390f3c20bda365ea11b50a0c9e10e3-s0623d4883330c6be11b-c1        143 2023-04-24 02:37:21.643 PDT []
xn185_5e921ef57e7f4b64c60308d2112f2ac8-s4f3bba61939835b311b-c1    1350882 2023-04-24 02:47:59.958 PDT []
xn185_496185448cd8847f603fd09afd1a5f43-s876f244f8144565811b-c1        143 2023-04-24 02:48:01.318 PDT []
xn185_fcc52594282ba2c6017b0a7f5a0e5607-s03b3d221db09ca6e11b-c1    1350882 2023-04-24 02:54:30.062 PDT []
xn185_d723b66526e5466b83e5296a9501064f-s07e9ffdf0a2b4f0911b-c1        143 2023-04-24 02:54:31.050 PDT []
xn185_f54c7ad94824480bac5c1a069742171a-s87234e6a3408d06811b-c1    1350950 2023-04-24 03:02:36.088 PDT []
xn185_9f21ed7952d83c926718ed5d99dce9a5-sa36f35ca5e189f5a11b-c1        143 2023-04-24 03:02:37.157 PDT []
xn185_df8d735c255e9708da443d828d2cb607-sb494565f19f25b7c11b-c1    1350950 2023-04-24 03:12:17.686 PDT []
xn185_9c074ce24016d6768247729cb3bbc40d-sc2432f93ade3e32e11b-c1       1615 2023-04-24 03:12:17.874 PDT []
xn185_b294c36ec9a28176b0bff931612001b1-s20c8e94ae0d1142c11b-c1        143 2023-04-24 03:12:18.910 PDT []
```

The superseded index is way too large, so this should mean that kopia has a lot of maintenance build up for when I'll run it next:

```
kopia index list --superseded | wc -l
2560
kopia index list | wc -l
52
```

Does this repro look like a good repro base to you guys ? I did not see anything wrong with kopia memory usage so far while running deletes and snapshots.

cc: @julio-lopez @ktogias @Lyndon-Li 

I'll run maintenance on this next and see if I am able to repro your scenario.

**Comment by vkamra** [comment]
> I'll run maintenance on this next and see if I am able to repro your scenario.

@Shrekster - what was useful when debugging https://github.com/kopia/kopia/pull/2956 was to generate a memory profile while running the repro (I think `--profile-memory` will do that). Would recommend doing that when running maintenance.

**Comment by Shrekster** [comment]
Thanks Vaibhav. I already ran without that flag and did not want to interrupt it. I will run again and can repro this again. I am planning to run this repro again, this time with the go-delve debugger (thanks to @julio-lopez for suggesting it) and also the `--profile-memory` flag.

Regardless, we might be onto something here, but the problem is not with index compaction it seems. Here is the maintenance run memory usage plot over time, out memory usage went upto 32GBs, definitely unexpected:

![plot](https://user-images.githubusercontent.com/1415030/234105646-e1ba9e29-b5ac-44b4-bb75-305dbe5b4494.svg)

There are three things going on here:

- index compaction (this happens outside of maintenance, based upon epochs)
- quick maintenance
- full maintenance

From the repo status we can tell that my repo has 185 epics and we must be running the heavy weight full index compaction every 7 epochs:
```
Epoch Manager:       enabled
Current Epoch: 185

Epoch refresh frequency: 3m0s
Epoch advance on:        20 blobs or 10 MiB, minimum 10m0s
Epoch cleanup margin:    4h0m0s
Epoch checkpoint every:  7 epochs
```

So, because I did not see any problem with memory usage during snapshots, that shows that index compaction is likely not the problem here.

@ktogias @Lyndon-Li I suspect that your kopia repo is actually running maintenance. From the logs you shared, it seems that the repo is scheduled for full maintenance every 90 days (you mentioned 2160 hours is the freq):

```
$ grep "full maintenance" /tmp/logs-from-velero-in-velero-84f99cc9dc-9wvxp.log
time="2023-04-24T13:12:05Z" level=debug msg="not due for full maintenance cycle until 2023-04-24 17:26:10.507876165 +0000 UTC" logModule=kopia/maintenance logSource="pkg/util/logging/kopia_log.go:44"
```

So, is it about 90 days since the creation of your repo now ? And any chance we could be running maintenance now ?

I also discussed my repro with @julio-lopez and he suggested a few more introspection as follows. In my repository I've a lot of deleted manifest contents & dirs (contents with k prefix & m prefix). The deleted contents with m prefix are likely snapshot roots:

```
kopia content list --deleted-only --prefix m | wc -l
2330

kopia content list --prefix k | wc -l
4374
```

@ktogias @Lyndon-Li can you also run the above two commands and paste their output ?

No indirect blobs, which makes sense because I have no large directories or files:

```
kopia content list --prefix x | wc -l
0
```

For full maintenance we do have to load all the dir contents. However, it is not clear to me why loading those json objects would shoot up the Kopia maintenance usage upto 32GBs of resident memory utilization. This is referring to the problem @ashmrtn wrote an improvement for https://github.com/kopia/kopia/pull/2956.

Next steps:

- We need the confirmation whether maintenance is the problem with the scenario @ktogias and @Lyndon-Li are also running into and that they have similar number of manifest/dir contents in their repo.
- I will load all the manifest/dirs in memory and check their memory usage to check their working in-memory utilization.
- Another very likely problem here could be that kopia is accumulating a lot of garbage in memory (either through json parsing or some other data structure). I will rerun the repro with go pprof profiles turned on this time. To answer this question better, I'll try to attach to the running process.
- I also want to run the repro with the current version of Kopia to see if there is any change in behavior.

**Comment by Shrekster** [comment]
As an add-on to my previous comment ...

I discussed the repro with @jkowalski. In my repro I did not have a lot of manifests, so not a whole lot of JSON parsing on that front:

```
$ kopia content list --prefix m -l
ma06b3f3b331dd9aa92727c81b2d401a9 99917 2023-04-24 03:12:17 PDT q6bd38ff17395aed038417745948d4d60-sc2432f93ade3e32e11b 32+99945 -
mb7429b521fde46bf48b3fcbb7c809aab 147 2023-04-24 03:12:18 PDT qabe472fe2acdbcc5ff04ecfb369e98a6-s20c8e94ae0d1142c11b 32+175 -
mfd4a46cd9a3cb6e660a4505bbc00bf1a 508 2023-04-24 03:12:17 PDT q1a10a88e19d0d30a87b06ef4dd964646-sb494565f19f25b7c11b 1144289+536 -
```

From what @ashmrtn brought up earlier, there were a lot of manifest here for her case.

My repro does have 4k dirs, which accounted for roughly 200MB of in-use allocations during my maintenance with memory profiling enabled.
![Screenshot 2023-04-24 at 3 13 38 PM](https://user-images.githubusercontent.com/1415030/234128011-6e0a949e-0aaf-4bec-b33c-a8244dd30399.png)

We might be aligning more towards the problem of maintenance now. This particular issue might just be about kopia maintenance not releasing the memory back to the OS as fast as it is producing garbage. As a generic statement, we might need to improve kopia maintenance to have a chunked approach instead. Still waiting on response on questions in the prior comment. I'll play with GOMEMLIMIT/GOGC to see if we can bring down the memory usage.

**Comment by ktogias** [comment]
Hi all, 

@Shrekster I thew following commands on my repo. Here are the results:

```
time docker exec kopia ./kopia snapshot list --all --password=<> | wc -l
101

real	7m0,998s
user	0m0,105s
sys	0m0,105s
```

```
time docker exec kopia ./kopia maintenance info --password=<>
Owner: default@default
Quick Cycle:
  scheduled: true
  interval: 1h0m0s
  next run: now
Full Cycle:
  scheduled: true
  interval: 24h0m0s
  next run: now
Log Retention:
  max count:       10000
  max age of logs: 720h0m0s
  max total size:  1 GiB
Recent Maintenance Runs:

real	5m31,688s
user	0m0,086s
sys	0m0,129s
```

```
time docker exec kopia ./kopia blob stats --password=<>
Got 10000 blobs...
Got 20000 blobs...
Got 30000 blobs...
Got 40000 blobs...
Got 50000 blobs...
Count: 56342
Total: 1.2 TB
Average: 20.7 MB
Histogram:

        0 between 0 B and 10 B (total 0 B)
        3 between 10 B and 100 B (total 54 B)
      123 between 100 B and 1 KB (total 52.4 KB)
      371 between 1 KB and 10 KB (total 1.6 MB)
       61 between 10 KB and 100 KB (total 3.2 MB)
      230 between 100 KB and 1 MB (total 124.5 MB)
     1084 between 1 MB and 10 MB (total 4.5 GB)
    54467 between 10 MB and 100 MB (total 1.2 TB)

real	20m12,379s
user	0m0,106s
sys	0m0,116s
```

```
time docker exec kopia ./kopia index list --password=<>
xs0_56cb8a340f35d629ffe22f4df77f5953-s28da87f8d88beae3-c1     274648016 2023-03-16 20:00:02.741 UTC []
xs1_8c9132863a466e65dded293f3353b919-sadffd632911df667-c1     198212231 2023-03-18 15:51:18.705 UTC []
xs2_fb3de18d6b7e88d458e0a2d73f5617dc-sb83795d242e03d40-c1         30070 2023-03-20 13:23:29.986 UTC []
xs3_0b8c366e57952055c7e5ef26035fa2c7-s6e63f6b8acc00d8a-c1      75176265 2023-03-21 13:55:12.241 UTC []
xs4_def2c5c740e3e5f0080ae7cd87cff422-s57bed458b7851ba1-c1      41263837 2023-03-22 14:42:35.274 UTC []
xs5_d2fa439a7445e77a47fb6205f262090c-sbd6163b72789d315-c1     161982039 2023-03-24 23:19:50.274 UTC []
xs6_f6ce050019668073036394d5a77a4657-s64f8c11b66f1d22c-c1      60328955 2023-03-27 23:30:41.784 UTC []
xn8_4ddb9e258540cd2498f7ae10b52930b7-s0a2c382e3328cf5411a-c1     137998 2023-03-27 23:41:04.846 UTC []
xn8_13e8bb87f00306d8094aebf5b37f026b-s6924d58f3ad6035b11a-c1      51547 2023-03-27 23:45:07.843 UTC []
xn8_93d9e297ef651aa3e1ba4b947ce68c4d-s9fa1f1d3caafbc0011a-c1      34676 2023-03-27 23:55:25.295 UTC []
xn8_53ef03ac7869cb5b0ac73d3b07fcadee-s1fc938ecc751894011a-c1     153450 2023-03-28 00:05:45.041 UTC []
xn8_eee984c1d2f0288998342d0fa460c65d-s115aca13aa4d24cc11a-c1       1297 2023-03-28 00:07:39.115 UTC []
xn8_c6ea98ddc20d1edab0b97306efdb5d20-scf6c681640926b3a11a-c1     174666 2023-03-28 23:17:42.150 UTC []
xn8_f79e195d1f667468958a63d4b5c31ebc-s856e8a2575f5e15e11a-c1      38305 2023-03-28 23:22:28.553 UTC []
xn8_d44bb6f9e9c49e02d18974b2d54437bd-sd63e3f0d4f15438211a-c1      35313 2023-03-28 23:32:41.863 UTC []
xn8_606180f3c5628930ddcd7380cc82fe44-s8ab66cb4b30fd4a211a-c1      16724 2023-03-28 23:42:52.307 UTC []
xn8_1f459fa5abe839df85675295cd3af837-sd9266dbabfe75efe11a-c1     161177 2023-03-28 23:51:06.015 UTC []
xn8_63f910b842630688a715e5b2c6db3605-sd8c69f2655c6be6f11a-c1     211835 2023-03-28 23:55:42.647 UTC []
xn8_169b650a6bf86dfc6e2de6c0f59477dd-sce09fa1f6652f7b411a-c1     211155 2023-03-29 00:00:14.783 UTC []
xn8_0442d9f1320e4c9835714c31f6d80b22-s1e0256280e23845b11a-c1     211817 2023-03-29 23:18:17.402 UTC []
xn8_e5fd2269ea1047e16457c7da738c2555-sfa9374f4b1234ae711a-c1       1025 2023-03-29 23:18:46.912 UTC []
xn8_93cd8b2bb4d6391ac54f00558bfab3a6-s67cb0008696343b811a-c1      85869 2023-03-29 23:28:56.355 UTC []
xn8_91904c45c2b1f885d0ad184585c47d6d-saa64f7ca5503c30211a-c1     126531 2023-03-29 23:39:29.804 UTC []
xn8_2220e674ace93665351a0059e1e71849-s8aadb51feeeeea5811a-c1        447 2023-03-29 23:40:12.730 UTC []
xn8_fd3d90a78ccd0b24a88768fde75e3394-s6881f083a854307611a-c1     211001 2023-03-29 23:50:37.223 UTC []
xn8_a635ec69e9bf515225ce552fd6ac896e-sce6a3601d592756811a-c1       1161 2023-03-29 23:50:57.916 UTC []
xn8_c876ddd4990b0e0f47b1c755038fe371-sb39000599f9256af11a-c1     211545 2023-03-29 23:55:48.504 UTC []
xs7_4ff5d750f1c1dd7e4ae680b1044422fd-s5130f4deab4b7127-c1       2488458 2023-03-30 00:28:14.194 UTC []

real	6m42,986s
user	0m0,087s
sys	0m0,103s
```

```
time docker exec kopia ./kopia index list --superseded --password=<> | wc -l
637

real	4m11,144s
user	0m0,069s
sys	0m0,107s
```

```
time docker exec kopia ./kopia index list --password=<> | wc -l
28

real	6m28,834s
user	0m0,125s
sys	0m0,140s
```

```
time docker exec kopia ./kopia repo status --password=<>
Config file:         /app/config/repository.config

Description:         Repository in S3: vouchernas.cti.gr:9000 vkub
Hostname:            a2533bf78576
Username:            nonroot
Read-only:           true
Format blob cache:   15m0s

Storage type:        s3
Storage capacity:    unbounded
Storage config:      {
                       "bucket": "vkub",
                       "prefix": "kopia/prod-minio01/",
                       "endpoint": "<>:9000",
                       "doNotUseTLS": true,
                       "accessKeyID": "minioadmin",
                       "secretAccessKey": "*********",
                       "sessionToken": ""
                     }

Unique ID:           c8ce5b5454f666947587938237afc7a8df3c595faf5b5b42b234ff0a5efafd9d
Hash:                BLAKE2B-256-128
Encryption:          AES256-GCM-HMAC-SHA256
Splitter:            DYNAMIC-4M-BUZHASH
Format version:      2
Content compression: true
Password changes:    true
Max pack length:     20 MiB
Index Format:        v2

Epoch Manager:       enabled
Current Epoch: 9

Epoch refresh frequency: 20m0s
Epoch advance on:        20 blobs or 10 MiB, minimum 24h0m0s
Epoch cleanup margin:    4h0m0s
Epoch checkpoint every:  7 epochs

real	7m21,566s
user	0m0,038s
sys	0m0,047s
```

```
time docker exec kopia ./kopia content list --deleted-only --prefix m --password=<> | wc -l
176

real	6m59,075s
user	0m0,033s
sys	0m0,033s
```

```
time docker exec kopia ./kopia content list --prefix k --password=<> | wc -l
13279161

real	5m1,458s
user	0m7,002s
sys	0m16,342s
```

I noticed that the actual output of the commands is produced quite quickly (usually 20-30 seconds) and then I have to wait for minutes with high cpu and memory usage for the command to complete and exit.

**Comment by ktogias** [comment]
Continuing my previous comment, this is the count of content list with x prefix:

```
time docker exec kopia ./kopia content list --prefix x --password=<> | wc -l
8444

real	4m56,459s
user	0m0,104s
sys	0m0,138s
```

**Comment by ktogias** [comment]
An here the content with m prefix:

```
time docker exec kopia ./kopia content list --prefix m -l --password=<>
m00f8b1dc0b9e05bc51ac31531bf215a7 809 2023-03-27 23:45:06 UTC q2713f680b963ca1e2e4fd9b2b1ac14bb-s6924d58f3ad6035b11a 901593+837 -
m0bf292d2ea3dc75dac7aa72ccab961f2 806 2023-03-29 23:18:45 UTC q4a96a8219f1bfb943140e3fbbd1b155d-sfa9374f4b1234ae711a 399780+834 -
m52169c47fa7febe8741b0ad5e4928022 809 2023-03-29 00:00:13 UTC qea4287e25a2f842f359b302f53fca68a-sce09fa1f6652f7b411a 1742169+837 -
m56c518b464a385b4a5c859fd4fdd9584 807 2023-03-29 23:40:12 UTC q7753a503eafeabd1d5a586db3e5d3784-s8aadb51feeeeea5811a 396953+835 -
m61b09d1302a2d5b24effa906bc7a4e67 811 2023-03-28 00:07:38 UTC qaf2be9e6bf195ed29c00b00ee48e239f-s115aca13aa4d24cc11a 400478+839 -
m7067821a4e1cdc698eba4d489612c3fc 808 2023-03-29 23:55:47 UTC q4b1406e4057151bc4ed682035cab7010-sb39000599f9256af11a 1767280+836 -
m742e332ed380e70d494e3b0faa3ce2e9 805 2023-03-28 23:55:41 UTC qf23b85ad918e34ce04d8bd7127970566-sd8c69f2655c6be6f11a 1743545+833 -
m84dffe0e5c2ccf69b80a1254736583e4 803 2023-03-28 23:51:02 UTC qba87d73e978ea456c626f3b7148267bb-sd9266dbabfe75efe11a 1479681+831 -
m984bf265033662b39d401d545f429b1e 12790 2023-03-27 00:09:21 UTC q6cf6fb8eb1e8e5f4d8ff29fd9005c604-sa91aec90649ecf7911a 32+12818 -
ma9ae0e9f69412de2b2ee8551ed192e3d 808 2023-03-27 23:16:44 UTC q3b74cbb8bd6d4af5395f94871e11a84f-s1ab939cfec382b3511a 1620899+836 -
md811e822d72e9b6b8ce4896525ea3a59 806 2023-03-27 23:30:15 UTC qedaa0f4dfb03bdd3ebdfd608305eb325-sc2166d74b43c9f6611a 688657+834 -
mf20a635db98a36a4d42119683e28e76a 805 2023-03-29 23:50:56 UTC q8be217ae346c75ea91dcecbd907c0515-sce6a3601d592756811a 400468+833 -
mf9f208f8864b47104c031906807a9204 807 2023-03-28 23:22:26 UTC q9d71ad0cd63c1be1d4ef56603ea9cd8e-s856e8a2575f5e15e11a 857387+835 -

real	6m55,452s
user	0m0,041s
sys	0m0,037s
```

**Comment by julio-lopez** [comment]
#1980 (680e3f29) was introduced after v0.10.7 (5d87d817)

**Comment by julio-lopez** [comment]
@ktogias Thanks for reporting the issue, and also with providing additional information to track things down.

@Shrekster had a repro and the data seems to point to the issue already resolved in #1980.
#1980 was done after `v0.10.7` as noted above. So the solution seems to be: _"please upgrade kopia"_

Shikhar, could you share some of the profile information that points to the allocation issue you observed?

Also, thanks to @jkowalski both for the help in understanding the issue and explaining the fix.

**Comment by julio-lopez** [comment]
Shikhar's work in reproducing the issue here and the resulting data also points out that we need to improve other aspects of maintenance with respect to resource consumption and scalability.

**Comment by ktogias** [comment]
@julio-lopez Please let me know when we have a velero build with upgraded kopia that I could test.

**Comment by Shrekster** [comment]
Here is the pprof dump from my repro:

[mem.pprof.gz](https://github.com/kopia/kopia/files/11390112/mem.pprof.gz)

![Screenshot 2023-05-03 at 12 11 05 PM](https://user-images.githubusercontent.com/1415030/236019885-55cb4a44-c126-409b-9317-80ec7c01c5a3.png)

The biggest block on allocations was index_v2.go:findEntry():

https://github.com/kopia/kopia/blob/5d87d817335f6d547e094ab80062113dc3a1fdf4/repo/content/index/index_v2.go#L373

Probably was that GoLang stack variables are not necessarily stack allocated when it cannot determine the variable usage outside of the local function stack memory. So, this was switched to use a fixed buffer reference in the PR above. (thanks to @jkowalski )

In my tests with latest Kopia the usage was cut approx in half. We might still be spending memory someone else, which we are still working upon. Also, to what Julio said, we need to rethink the design on maintenance in the long term. Having said that, the latest Kopia does get benefited a lot by the above PR, so my recommendation too is to upgrade your Kopia version.

Here is the memory usage plot from the latest Kopia on my exact same repo which I created using `v0.10.7` above.

![plot](https://user-images.githubusercontent.com/1415030/236022356-973dd084-7b71-462f-8c0e-be568d134922.svg)

I'll replay this with profiling enabled again, but lets take this second part of the investigation off this issue.

**Comment by Lyndon-Li** [comment]
@Shrekster @julio-lopez
Thanks a lot for investigating this issue, so looks like we've found two problems in v0.10.7 which could both be solved by upgrading.

@ktogias
For now, please test with Kopia v0.13 directly, and let us know if the problem is solved.
Velero is working on Kopia bumping up, so in the coming releases, it should integrate with the latest version of Kopia.

**Comment by ktogias** [comment]
@Lyndon-Li 

I run all the above commands using kopia/kopia:0.13 docker image and I confirm that I do not observe high memory usage. Memory and cpu usage are normal, the system remains stable with out limiting container memory and all commands complete in much less time than with 0.10 image.

**Comment by Lyndon-Li** [comment]
@ktogias Thanks for the information, then please wait some time for the next patch release of Velero.

**Comment by julio-lopez** [comment]
> @julio-lopez Please let me know when we have a velero build with upgraded kopia that I could test.

That's up to the Velero folks. https://github.com/kopia/kopia/issues/2888#issuecomment-1534327075

**Comment by julio-lopez** [comment]
Closing this as resolved.

Thanks @ktogias for all the info on behalf of the kopia project.
